internal/controller/supervisorconfig/oidcupstreamwatcher/oidc_upstream_watcher.go

```diff
@@ -351,6 +351,44 @@ func (c *oidcWatcherController) validateIssuer(ctx context.Context, upstream *v1
 		c.validatorCache.putProvider(&upstream.Spec, discoveredProvider, httpClient)
 	}
 
+	// Get the revocation endpoint, if there is one. Many providers do not offer a revocation endpoint.
+	var additionalDiscoveryClaims struct {
+		// "revocation_endpoint" is specified by https://datatracker.ietf.org/doc/html/rfc8414#section-2
+		RevocationEndpoint string `json:"revocation_endpoint"`
+	}
+	if err := discoveredProvider.Claims(&additionalDiscoveryClaims); err != nil {
+		// This shouldn't actually happen because the above call to NewProvider() would have already returned this error.
+		return &v1alpha1.Condition{
+			Type:    typeOIDCDiscoverySucceeded,
+			Status:  v1alpha1.ConditionFalse,
+			Reason:  reasonInvalidResponse,
+			Message: fmt.Sprintf("failed to unmarshal OIDC discovery response from %q:\n%s", upstream.Spec.Issuer, truncateMostLongErr(err)),
+		}
+	}
+	if additionalDiscoveryClaims.RevocationEndpoint != "" {
+		// Found a revocation URL. Try to parse it.
+		revocationURL, err := url.Parse(additionalDiscoveryClaims.RevocationEndpoint)
+		if err != nil {
+			return &v1alpha1.Condition{
```

Comment:
I think that by returning this condition we automatically fail hard and prevent the user from logging in completely? If so, this feels a little harsh.

 If everything else is correct, but, say the revocation endpoint is http, not https, then that's a problem and we shouldn't try to revoke your tokens over plaintext connection. _But_ this whole feature seems like just a courtesy, to let the idp know when we're done, and if it doesn't work out I don't know if it's worth preventing login over. Could we report it as a condition but still keep going as if you didn't have a revocation endpoint?

Reply:
That's a really good catch, @margocrawf. I actually had not thought of that and it is definitely worth considering.

Note that RFC 7009 specifies:

> "Since requests to the token revocation endpoint result in the transmission of plaintext credentials in the HTTP request, URLs for token revocation endpoints MUST be HTTPS URLs.  The authorization server MUST use Transport Layer Security (TLS) [RFC5246] in a version compliant with [RFC6749], Section 1.6."
>
> From: https://datatracker.ietf.org/doc/html/rfc7009#section-2

So there should not be any implementations in the wild which advertise an invalid URL or a non-TLS URL for token revocation. If they did, they would be in blatant contradiction of the spec. That makes me feel like we're probably okay to consider that an error for now. What do you think?

Reply:
@cfryanr thanks, that's useful context. It would definitely indicate a misconfiguration on their part, and I hope that none of this is ever relevant because people are doing the right thing with their revocation endpoints 😆...
I'm still not convinced that it's a misconfiguration _worth failing over_, because I don't see it as a big deal if we aren't able to use it.
But I think it's probably fine either way because it's such an edge case.